### ISSUE TITLE
Inappropriate error code when reading a request body fails

### ISSUE THREAD

**Issue by svanimpe** [IBM-Swift/Kitura#1271]
## Context and Description

In the following example, when reading the `Message` fails due to a misspelled property, Kitura returns a 404 response:

```swift
struct Message: Codable {
    
    let text: String
}

router.post("message") {
    request, response, next in
    let newMessage = try request.read(as: Message.self)
    message = newMessage
    response.send(json: message)
    next()
}
```

Is it possible to return something more informative, like 400 or 422? A 404 is really confusing here, as the route does exists.

I currently use the following workaround:

```swift
guard let newMessage = try? request.read(as: Message.self) else {
    response.status(.badRequest)
    return next()
}
```

### Environment Details

Kitura 2.3, Swift 4.1 on macOS.

**Comment by ianpartridge** [comment]
@Andrew-Lees11 could you comment on this?

@svanimpe my immediate reaction is that your workaround is actually the recommended way to do this. 404 is the default error response from "request, response next" routes in Kitura - if you want to change this you can either set `response.status()` in your route, or inside a `router.error()` handler.

**Comment by Andrew-Lees11** [comment]
@svanimpe
I agree with Ian that your current handling of the error is the correct way to set the response status.
When you call the read(as:) function it doesn't have access to the response and so cannot set the error. 
The uncaught throw is then handled by the "request, response next" route and defaults to 404.

**Comment by svanimpe** [comment]
Ah, I guess I assumed the error contained information based on which the handler could choose an appropriate code.

Isn't that what happens with codable routes, where you return an error in the completion handler?
